README.rst
A Last.fm API interface.

Features
========

- Supports all packages and methods listed at http://www.last.fm/api.

- Methods return plain Python types: dictionaries and lists.

- If the user wishes, the library can make an application comply to the point 4.4 of
  `Last.fm's API ToS <http://www.last.fm/api/tos>`_, which says that the request
  rate of an application must be limited.

- Handles all Last.fm API errors via exceptions.

- Easily extendible.

|

A tutorial
==========

A short introduction
--------------------

*lfm.App*
~~~~~~~~~

Instantiate a Last.fm application::

    from lfm import lfm
    
    app = lfm.App(API_KEY, SECRET)

The above is self-explanatory. You'll need an API key and the corresponding "secret"
given by Last.fm. If you don't have those handy, you can ommit them for **testing**
purposes, as *lfm* comes with its own. You are expected to provide your own key and
secret in real applications.

If you want your application to comply to Last.fm's request rate limit, you'll need
to provide a third argument, a file in which a sqlite3 database which tracks requests
will be stored.

::
    
    LFM_FILE = "lfm.dat"
    
    app = lfm.App(API_KEY, SECRET, LFM_FILE)

As a fourth argument, you can provide a tuple of your program's name and version,
to be used in the user-agent::

    NAME    = "myprogram"
    VERSION = "1.0.0"
    
    app = lfm.App(API_KEY, SECRET, LFM_FILE, (NAME, VERSION))
    
The user-agent is formatted as "NAME/VERSION lfm/LFMVERSION". If you don't provide
this information, both the name and version will be "unknown".


Methods and packages
~~~~~~~~~~~~~~~~~~~~

API methods are organized like so::

    data = app.package.method_name(...)
    
So, if you wanted, for example, to fetch all recently listened tracks of a user,
you'd do something like this::

    tracks = app.user.get_recent_tracks(user)
    
Note the underscores. Last.fm uses *camelCase* for method names. Such a thing
isn't Pythonic, though, hence the transformation of names to *under_scores*.


Authenticating
--------------

*auth.get_mobile_session()*
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Let's get a user's session now. There are three ways to do this. The first one
is by supplying a username and a password::

    session = app.auth.get_mobile_session(user, pwd)


*auth.get_session()*
~~~~~~~~~~~~~~~~~~~~
    
The second one is more complicated, but more secure and trustworthy. First,
you need to fetch a token::

    token = app.auth.get_token()
    
Then you have to make the user authenticate the token by pointing him to the
authentication web-page::

    import webbrowser
    
    webbrowser.open(token.url)
    input("Press enter after granting access.")
    
After the user has granted access, all that's left is to fetch the session::

    session = app.auth.get_session(token)


*auth.get_url()*
~~~~~~~~~~~~~~~~~~~~

The third one is for web apps. `auth.get_url()` takes one argument, a callback url,
and returns an url which points to the authentication page. When the user
authenticates, the page redirects to the callback url with the token appended as a
GET parameter.  
For more information, consult the `official Last.fm documentation <http://www.last.fm/api/webauth>`_.


Using the session
~~~~~~~~~~~~~~~~~

Regardless of which of the methods you use, a session needs to be bound to your
app by assigning the session key to the *App*'s *session_key* attribute::

    app.session_key = session["key"]

That's all. You can now call methods which require authentication::

    app.track.remove_tag(artist, track, tag)
    

More
----

For more information on specific methods, consult the `API page <http://www.last.fm/api>`_
on Last.fm.

|

An advanced tutorial
====================

Custom requests
---------------

Say, for example, that Last.fm has added a new method not yet available in
this library. What can be done then? The solution is actually quite straightforward:
use *App.request()*. You can manually specify the API package, method and parameters::

    def playlist_remove(app, playlist_id):
        params = {
                  "playlistID": playlist_id,
                 }
    
        return app.request("playlist", "remove", params)

Simple as that.


Adding new packages
-------------------

*App.request_auto()*
~~~~~~~~~~~~~~~~~~~~~~~~

What if, by some miracle, a whole new package with a bunch of methods was added?
You'd want to use those methods several times in your program. Calling *request()*
every time would be quite cumbersome; very repetitive and error-prone.

Well, an unlikely hero arises: *App.request_auto()*! This function tries
to automate every bit of requesting that can possibly be automated, and generally
succeeds very well! This whole library is built on that one function. Here's an
example from the source itself::

    class Track(Package):
    
        ...
        
        def get_info(self, artist = None, track = None, username = None, autocorrect = None, mbid = None):
            data = self.app.request_auto()
            return data["track"]
        
        ...

What kind of magick is this? Well, without going into too much detail
(open source, remember?), the function cleverly learns all of the three,
if possible: the package, the method, the parameters:

- It assembles the method name from the caller function's name; "getInfo"
  in this case.
  
- The parameters, ignoring *self*, are grabbed from the caller's arguments.
  Parameter names are stripped of trailing underscores, to allow the use of
  parameters such as *from*.
  True to the Python's philosophy of "duck-tape" programming, the function tries
  to accept all kinds of types as parameters. It handles all primitive ones well:
  integers, floats, booleans, and such. Of the more complicated types, it can
  handle lists, but not dictionaries.
  
- The name of the package is learned from the name of the class the function's
  in, but **only** if the class inherits *lfm.Package*.
  
*request_auto()* is not only intelligent and elegant, it's also flexible.
You can override any of the three::

    def get_info(self, artist = None, track = None, username = None, autocorrect = None, mbid = None):
        package = "the_correct_package_name"
        method  = "the_correct_method_name"
        
        params  = {
                   "special" : 0xDEADBEEF,
                   "mbid"    : None,
                  }
        
        data = self.app.request_auto(package, method, params)
        return data["track"]
        
So, we have added a new parameter called *special*, and made *mbid*
always *None*, whatever the user may have passed. Pretty neat, huh?
Note that *params* will be **merged into** the auto-gathered
dictionary of parameters, not overwrite them. 


Inheriting *Package*
~~~~~~~~~~~~~~~~~~~~

Very well, your custom-made Package would look something like this::

    class Forum(Package):
        def post(self, threadid, msg):
            data = self.app.request_auto()
            return data
    
And you'd use it like so::

    forum = Forum(app)
    forum.post("1832723", "Hello folks!")


Inheriting *App*
~~~~~~~~~~~~~~~~

To add the finishing touch, you could extend *App*::

    class App(lfm.App):
        forum = None
        
        def __init__(self, key, secret, db = None, info = None):
            super().__init__(key, secret, db, info)
            
            forum = Forum(self)

And with that::

    app.forum.post("1832723", "Hello folks!")
    


CHANGES.txt
v1.1.1, 2014-04-17:
 - Fixed a bug in auth.get_token().

v1.1.0, 2014-03-30:
 - Changed the directory structure to a simpler one, breaking backwards compatibility, and fixing some problems.
   Practically speaking, you'll now do 'from lastfm import lfm' as opposed to just 'import lfm'.
 - Renamed 'app.sk' to 'app.session_key'.

v1.0.5, 2014-03-30:
 - Added an authentication method for web apps: app.auth.get_url(callback)
 - Fixed info.py being in the global namespace

v1.0.4, 2014-03-29
 - Added a constant: 'MAX_USERNAME_LENGTH'.
 - Applied various minor fixes.

v1.0.3, 2013-07-29
 - Cleaned up the code related to logging of requests.
 - Added a user-agent, and a user-agent option.
 - The package now imports everything from lfm.lfm.
 - Better code organization here and there.

v1.0.2, 2013-07-25
 - Fixed tag.get_info() having "artist" and "mbid" parameters instead of just "tag".
 - Fixed a bug where requests would be denied because of capitalized package names.
 - Fixed a bug in a utility function where only keys of a dictionary were being evaluated,
   instead of the values as well.

v1.0.1, 2013-07-23
 - A shaped, working release.

v1.0.0, 2013-06-26
 - The initial release.
 


lfm/api.py
import requests

from abc import ABCMeta, abstractmethod
from functools import wraps
import inspect
import xml.etree.ElementTree as XmlElementTree
import hashlib

import lfm.info as info


USER_AGENT  = "{}/{}".format(info.NAME, info.VERSION)

ROOT                    = "https://ws.audioscrobbler.com/2.0/"
REQUEST_RATE_PERIOD     = 300
REQUEST_RATE_INTERVAL   = 5
MAX_REQUESTS            = REQUEST_RATE_PERIOD / REQUEST_RATE_INTERVAL
MAX_USERNAME_LENGTH     = 15

KEY     = "312a23775128fd0f9a6d8d3e7a87a4b4"
SECRET  = "d03bece4e91f9a25b0b2b78c75c7c327"


# Errors as shown across Last.fm API pages
class RequestError(Exception): pass
class InvalidServiceError(RequestError): pass
class InvalidMethodError(RequestError): pass
class AuthenticationFailedError(RequestError): pass
class InvalidFormatError(RequestError): pass
class InvalidParametersError(RequestError): pass
class InvalidResourceSpecifiedError(RequestError): pass
class OperationFailedError(RequestError): pass
class InvalidSessionKeyError(RequestError): pass
class InvalidApiKeyError(RequestError): pass
class ServiceOfflineError(RequestError): pass
class SubscribersOnlyError(RequestError): pass
class InvalidMethodSignatureError(RequestError): pass
class TokenNotAuthorizedError(RequestError): pass
class TokenExpiredError(RequestError): pass
class TemporaryErrorError(RequestError): pass
class TrialExpiredError(RequestError): pass
class NotEnoughContentError(RequestError): pass
class NotEnoughMembersError(RequestError): pass
class NotEnoughFansError(RequestError): pass
class NotEnoughNeighboursError(RequestError): pass
class SuspendedApiKeyError(RequestError): pass
class StationDeprecatedError(RequestError): pass
class GeoRestrictedError(RequestError): pass
class RateLimitExceededError(RequestError): pass

error_codes = {
    2: InvalidServiceError,              # 2 : Invalid service - This service does not exist
    3: InvalidMethodError,               # 3 : Invalid Method - No method with that name in this package
    4: AuthenticationFailedError,        # 4 : Authentication Failed - You do not have permissions to access the service
    5: InvalidFormatError,               # 5 : Invalid format - This service doesn't exist in that format
    6: InvalidParametersError,           # 6 : Invalid parameters - Your request is missing a required parameter
    7: InvalidResourceSpecifiedError,    # 7 : Invalid resource specified
    8: OperationFailedError,             # 8 : Operation failed - Something else went wrong
    9: InvalidSessionKeyError,           # 9 : Invalid session key - Please re-authenticate
    10: InvalidApiKeyError,              # 10 : Invalid API key - You must be granted a valid key by last.fm
    11: ServiceOfflineError,             # 11 : Service Offline - This service is temporarily offline. Try again later.
    12: SubscribersOnlyError,            # 12 : Subscribers Only - This station is only available to paid last.fm subscribers
    13: InvalidMethodSignatureError,     # 13 : Invalid method signature supplied
    14: TokenNotAuthorizedError,         # 14 : This token has not been authorized
    15: TokenExpiredError,               # 15 : This token has expired
    16: TemporaryErrorError,             # 16 : There was a temporary error processing your request. Please try again
    18: TrialExpiredError,               # 18 : Trial Expired - This user has no free radio plays left. Subscription required
    20: NotEnoughContentError,           # 20 : Not Enough Content - There is not enough content to play this station
    21: NotEnoughMembersError,           # 21 : Not Enough Members - This group does not have enough members for radio
    22: NotEnoughFansError,              # 22 : Not Enough Fans - This artist does not have enough fans for for radio
    23: NotEnoughNeighboursError,        # 23 : Not Enough Neighbours - There are not enough neighbours for radio
    26: SuspendedApiKeyError,            # 26 : Suspended API key - Access for your account has been suspended, please contact Last.fm
    27: StationDeprecatedError,          # 27 : Deprecated - This station is no longer available
    28: GeoRestrictedError,              # 28 : Geo Restricted - This station is not available with this client/country combination
    29: RateLimitExceededError,          # 29 : Rate limit exceeded - Your IP has made too many requests in a short period
    }


def request(key, secret, package, method, params = {}, session = None, name = None, version = None):
    params = forge_params(params, key, secret, package, method, session)
    resp = requests.post(ROOT, params, headers = forge_headers(name, version)).text
    
    return resp
    
    
def forge_params(params, key, secret, package, method, session = None):
    params["api_key"] = key
    params["method"] = "{}.{}".format(package, method)
    try: params["sk"] = session.key
    except AttributeError: pass
    params["api_sig"] = sign_params(params, secret)
    
    return params
    
    
def sign_params(params, secret):
    # Parameters are alphabetically sorted by their key, and then concatenated
    # in a keyvalue manner. The application's secret is appended afterwards,
    # the whole thing is UTF-8 encoded, and then md5() hashed, hex digest of it
    # being the signature. Some parameters mustn't be included in the calculation.

    # Keys excluded from the calculation
    excluded_keys = ["format"]

    concat_params = ""
    for key in sorted(list(params)):
        if key not in excluded_keys:
            concat_params += key + params[key]
    concat_params += secret
    
    sig = hashlib.md5(concat_params.encode("utf-8")).hexdigest()
    
    return sig
    
    
def forge_headers(name, version):
    user_agent = "{}/{} {}".format(name, version, USER_AGENT)
    return {
        "User-Agent": user_agent,
        }
        
        
def response_to_xml(response):
    return XmlElementTree.fromstring(response)
    
    
def check_xml_for_errors(xml):
    if xml.attrib["status"] == "failed":
        error = xml.find("error")
        error_code = int(error.attrib["code"])
        message = error.text.strip()
        raise error_codes[error_code](message)
    

class App:
    def __init__(self, key = KEY, secret = SECRET, session = None, name = "unknown", version = "unknown"):
        self.key = key
        self.secret = secret
        self.session = session
        self.name = name
        self.version = version
        
        
    def request(self, package, method, params = {}):
        return request(self.key, self.secret, package, method, params,  \
                        session = self.session, name = self.name,   \
                        version = self.version)


class PackageMetaclass(type):
    def __init__(cls, name, bases, attrs):
        api_package_name = name.lower()
        api_methods = [(k, v) for k, v in cls.__dict__.items()  \
                        if not k.startswith("__") and callable(v) and not hasattr(v, "_is_helper")]
                        
        for api_method_func_name, api_method in api_methods:
            try:
                api_method_name = api_method._name
            except AttributeError:
                api_method_name = api_method_func_name.replace("_", "").lower()
                
            api_method_params = [x.rstrip('_') for x in inspect.signature(api_method).parameters if x != "self"]
            
            try:
                api_method_returns = api_method._returns
            except AttributeError:
                api_method_returns = None
            
            @wraps(api_method)
            def wrapper(api_method, api_method_name, api_method_params, api_method_returns):
                @wraps(wrapper)
                def wrapper2(self, *args, **kwargs):
                    params = dict(zip(api_method_params, args))
                    
                    params_renames = api_method(self, *args, **kwargs)
                    if params_renames is not None:
                        for rename in params_renames.items():
                            params[rename[1]] = params.pop(rename[0])
                    
                    for k, v in params.copy().items():
                        if isinstance(v, bool):
                            if v: params[k] = "1"
                            else: params[k] = "0"
                        else:
                            try: param = v.as_parameter(k)
                            except AttributeError:
                                params[k] = str(v)
                            else:
                                params[k] = None
                                params.update(param)
                    
                    params = dict((k, v) for k, v in params.items() if v is not None)
                    
                    response = self.app.request(api_package_name, api_method_name, params)
                    if api_method_returns is not None:
                        return api_method_returns.from_xml(response, self.app)
                    else:
                        return response
                    
                return wrapper2

            w = wrapper(api_method, api_method_name, api_method_params, api_method_returns)
            setattr(cls, api_method_func_name, w)
            
        return super().__init__(name, bases, attrs)

        
class Package(metaclass = PackageMetaclass):
    def __init__(self, app):
        self.app = app
    

def name(which):
    def decorator(func):
        func._name = which
        return func
    
    return decorator


def returns(what):
    def decorator(func):
        func._returns = what
        return func
    
    return decorator
    

def helper(func):
    func._is_helper = True
    return func
    

class DataMetaclass(ABCMeta):
    def __new__(cls, name, bases, attrs):
        return super().__new__(cls, name, bases, attrs)

    def __init__(cls, name, bases, attrs):
        try:
            from_xml = cls.from_xml
        except AttributeError: pass
        else:
            @wraps(from_xml)
            def wrapper(response, app):
                xml = response_to_xml(response)
                check_xml_for_errors(xml)
                return from_xml(xml, app)
                
            cls.from_xml = wrapper

        return super().__init__(name, bases, attrs)
    
    
class Data(metaclass = DataMetaclass):
    @staticmethod
    def from_xml(xml, app):
        pass

    @abstractmethod
    def as_parameter(self, key):
        pass
        